internal/provider/validate.go

```diff
@@ -4,269 +4,355 @@
+			options.FrontMatter = LegacyGuideFrontMatterOptions
+		} else {
+			options.FrontMatter = LegacyFrontMatterOptions
+		}
+		v.logger.infof("running file checks on %s", rel)
+		result = errors.Join(result, check.NewProviderFileCheck(options).Run(path))
+
+		files = append(files, path)
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("error walking directory %q: %w", dir, err)
+	}
+
+	mismatchOpt := &check.FileMismatchOptions{
+		ProviderShortName: providerShortName(v.providerName),
+		Schema:            v.providerSchema,
 	}
+
+	if dirExists(filepath.Join(dir, "d")) {
+		dataSourceFiles, _ := os.ReadDir(filepath.Join(dir, "d"))
+		mismatchOpt.DatasourceEntries = dataSourceFiles
+	}
+	if dirExists(filepath.Join(dir, "r")) {
+		resourceFiles, _ := os.ReadDir(filepath.Join(dir, "r"))
+		mismatchOpt.ResourceEntries = resourceFiles
+	}
+	if dirExists(filepath.Join(dir, "functions")) {
+		functionFiles, _ := os.ReadDir(filepath.Join(dir, "functions"))
+		mismatchOpt.FunctionEntries = functionFiles
+	}
+
+	v.logger.infof("running file mismatch check")
+	if err := check.NewFileMismatchCheck(mismatchOpt).Run(); err != nil {
+		result = errors.Join(result, err)
+	}
+
+	return result
 }
 
-func checkAllowedFiles(dirs ...string) check {
-	allowedFiles := map[string]bool{}
-	for _, d := range dirs {
-		allowedFiles[d] = true
+func dirExists(name string) bool {
+	if _, err := os.Stat(name); err != nil {
+		return false
 	}
 
-	return func(dir string) ([]issue, error) {
-		issues := []issue{}
+	return true
+}
+
+var LegacyFrontMatterOptions = &check.FrontMatterOptions{
+	NoSidebarCurrent:   true,
+	RequireDescription: true,
+	RequireLayout:      true,
+	RequirePageTitle:   true,
+}
 
-		f, err := os.Open(dir)
-		if err != nil {
-			return nil, err
-		}
-		infos, err := f.Readdir(-1)
-		if err != nil {
-			return nil, err
-		}
+var LegacyIndexFrontMatterOptions = &check.FrontMatterOptions{
+	NoSidebarCurrent:   true,
+	NoSubcategory:      true,
+	RequireDescription: true,
+	RequireLayout:      true,
+	RequirePageTitle:   true,
+}
 
-		for _, fi := range infos {
-			if fi.IsDir() {
-				continue
-			}
+var LegacyGuideFrontMatterOptions = &check.FrontMatterOptions{
+	NoSidebarCurrent:   true,
+	RequireDescription: true,
+	RequireLayout:      true,
+	RequirePageTitle:   true,
+}
 
-			if !allowedFiles[fi.Name()] {
-				issues = append(issues, issue{
-					file:    filepath.Join(dir, fi.Name()),
-					message: fmt.Sprintf("file %q is not allowed", fi.Name()),
-				})
-			}
-		}
+var RegistryFrontMatterOptions = &check.FrontMatterOptions{
+	NoLayout:         true,
+	NoSidebarCurrent: true,
+}
 
-		return issues, nil
-	}
+var RegistryIndexFrontMatterOptions = &check.FrontMatterOptions{
+	NoLayout:         true,
+	NoSidebarCurrent: true,
+	NoSubcategory:    true,
+}
+
+var RegistryGuideFrontMatterOptions = &check.FrontMatterOptions{
+	NoLayout:         true,
+	NoSidebarCurrent: true,
+	RequirePageTitle: true,
```

Comment:
super nit 🦸🏻: Typically global package variables are declared at the top of the file